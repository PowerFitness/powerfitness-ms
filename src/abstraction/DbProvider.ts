import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
