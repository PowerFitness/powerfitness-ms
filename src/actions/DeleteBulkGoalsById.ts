import { Action } from '../abstraction/Action';
import DbProvider from '../abstraction/DbProvider';
import { OkPacket } from 'mysql';
import { buildWhereOrConditions } from '../util/buildConditions';

export class DeleteBulkGoalsById extends Action<OkPacket> {
	goalIds: Array<number>
	constructor(dbProvider: DbProvider, goalIds: Array<number> = []) {
		super(dbProvider);
		this.goalIds = goalIds;
	}
	buildQuery(): string {
		if (!this.goalIds.length) throw new Error('Missing criteria')
		return `delete from goal where ${buildWhereOrConditions(this.goalIds, 'id')}`;
	}
}

export default DeleteBulkGoalsById;
