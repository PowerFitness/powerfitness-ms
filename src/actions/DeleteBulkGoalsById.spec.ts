import DbProvider from '../abstraction/DbProvider';
import { mock } from 'jest-mock-extended';
import DeleteBulkGoalsById from './DeleteBulkGoalsById';

describe('DeleteBulkGoalsById', () => {
	test('buildQuery', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: DeleteBulkGoalsById = new DeleteBulkGoalsById(dbProvider, [ 1, 2 ]);
		expect(action.buildQuery()).toMatchSnapshot();
	})
	test('buildQuery throws an error if no criteria', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: DeleteBulkGoalsById = new DeleteBulkGoalsById(dbProvider, undefined as unknown as Array<number>);
		expect(() => action.buildQuery()).toThrowError('Missing criteria')
	})
})
