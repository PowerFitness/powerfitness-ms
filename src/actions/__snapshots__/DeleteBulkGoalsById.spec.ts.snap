// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteBulkGoalsById buildQuery 1`] = `"delete from goal where id = 1 OR id = 2"`;
